cmake/helper_functions.cmake

```diff
@@ -12,6 +12,14 @@ function(CREATE_EAGER_TEST_EXE TESTLIST)
             set(TEST_TARGET ${TEST_NAME})
         endif()
         add_executable(${TEST_TARGET} ${TEST_SRC_PATH})
+        set_target_properties(
+            ${TEST_TARGET}
+            PROPERTIES
+                UNITY_BUILD
+                    ON
```

Comment:
From the [official docs](https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD.html):

> Note Projects should not directly set the UNITY_BUILD property or its associated [CMAKE_UNITY_BUILD](https://cmake.org/cmake/help/latest/variable/CMAKE_UNITY_BUILD.html#variable:CMAKE_UNITY_BUILD) variable to true. Depending on the capabilities of the build machine and compiler used, it might or might not be appropriate to enable unity builds. Therefore, this feature should be under developer control, which would normally be through the developer choosing whether or not to set the [CMAKE_UNITY_BUILD](https://cmake.org/cmake/help/latest/variable/CMAKE_UNITY_BUILD.html#variable:CMAKE_UNITY_BUILD) variable on the [cmake(1)](https://cmake.org/cmake/help/latest/manual/cmake.1.html#manual:cmake(1)) command line or some other equivalent method. However, it IS recommended to set the UNITY_BUILD target property to false if it is known that enabling unity builds for the target can lead to problems.

I suspect [one reason](https://cmake.org/cmake/help/latest/variable/CMAKE_UNITY_BUILD.html) for this is:

> Note This option currently does not work well in combination with the [CMAKE_EXPORT_COMPILE_COMMANDS](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html#variable:CMAKE_EXPORT_COMPILE_COMMANDS) variable.

And the compile commands export is necessary for pretty much every piece of tooling that needs to understand C++ (eg: clang-tidy, iwyu, etc).

Clearly we have much work to do for unity builds to work on much of our code, so a dev enabling Unity globally is going to be met with sadness.  Probably best to introduce our own variable and light up targets based on that var, so we don't need to comb through every existing target and say "Never do Unity (until we come back around and fix)".

Reply:
Should we just add `-DCMAKE_UNITY_BUILD=ON` as a convenience thing to `build_metal.sh` with some helpful comments, based on CMake's official docs then? Or am I reading this wrong?

Reply:
Yes, but first we'd need to set `UNIT_BUILD=FALSE` on all the targets that break with unity builds (ie: everything not covered in this PR).
We probably should/must also set `CMAKE_UNITY_BUILD=FALSE` in `dependencies.cmake` so that we don't try doing unity builds on 3rd party stuff we don't have control over.

Reply:
Unity builds is something which must be kept opt-in for targets. It can impact build times both ways and break builds even for valid C++ code.
I made an option to disable unity builds for our targets to cmake and `build_metal.sh`, and auto-disabled it if `CMAKE_EXPORT_COMPILE_COMMANDS` is on.
cc @afuller-TT @tt-rkim

Reply:
So just a couple targets, eh? I count 4? 5? Good for me

Reply:
Typically you would turn it on for targets which have a lot of files and each of them has a lot of slow-to-compile includes. I think this PR enables unity build for all such targets, except for `tt_metal`.  In practice turning it on requires you to go and fix code inside of the target, so it's always a balance if it's worth it.